### PR TITLE
goldwarden: init at 0.2.7

### LINKED ISSUE #278180: Package request: goldwarden
**Project description**

Goldwarden is a Bitwarden compatible desktop client. It focuses on providing useful desktop features that the official tools do not (yet) have or are not willing to add, and enhanced security measures that other tools do not provide, such as:

    Support for SSH Agent (Git signing and SSH login)
    System wide autotype (Gnome, KDE only for now)
    Biometric authentication (via Polkit) for each credential access
    Implements Bitwarden browser-extension biometrics on Linux
    Support for injecting environment variables into the environment of a cli command
    Vault content is held encrypted in memory and only briefly decrypted when needed
    Kernel level memory protection for keys (via the memguard library)
    Additional measures to protect against memory dumps
    Passwordless login (Both logging in, and approving logins)
    Fido2 (Webauthn) support
    more to come...?

The aim is not to replace the official clients, but to complement by implementing the missing features.

**Metadata**

* homepage URL: https://github.com/quexten/goldwarden
* source URL: https://github.com/quexten/goldwarden
* license: mit
* platforms: linux, darwin

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by arthsmn**
## Description of changes

Homepage: https://github.com/quexten/goldwarden
Closes: https://github.com/NixOS/nixpkgs/issues/278180

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by arthsmn** [comment]
@quexten I looked at the aur package and it only seems to install a binary and a polkit rule, that's what is being done here. Are there any other steps needed in the package?

**Comment by quexten** [comment]
If selinux is present (does nix have it?), the polkit rule's permissions need to be adjusted as well. Other than that no, it's just installing the binary.

Do beware that you do need at least dbus, and a pinentry client (https://pkg.go.dev/github.com/twpayne/go-pinentry#section-readme) are required to use the package. I'm not sure if these are present on a default nixos installation.

**Comment by quexten** [comment]
Oh, since I do see darwin listed above, I will note that while packages are built for darwin in the project's GitHub CI at the moment, they are very much untested and some features are missing. I'm not sure they should be released for darwin just yet until more testing has been done there. https://github.com/quexten/goldwarden/issues/4

**Comment by arthsmn** [comment]
> Oh, since I do see darwin listed above, I will note that while packages are built for darwin in the project's GitHub CI at the moment, they are very much untested and some features are missing. I'm not sure they should be released for darwin just yet until more testing has been done there. [quexten/goldwarden#4](https://github.com/quexten/goldwarden/issues/4)

Understood, I'll mark it as linux-only and link the issue.

**Comment by arthsmn** [comment]
> If selinux is present (does nix have it?), the polkit rule's permissions need to be adjusted as well. Other than that no, it's just installing the binary.
> 
> Do beware that you do need at least dbus, and a pinentry client (https://pkg.go.dev/github.com/twpayne/go-pinentry#section-readme) are required to use the package. I'm not sure if these are present on a default nixos installation.

No, we currently don't support SELinux.

I added the packages to the binary's path, so they'll always be there.

File: pkgs/by-name/go/goldwarden/package.nix [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  nativeBuildInputs = [makeWrapper];
+
+  buildInputs = [libfido2];
+
+  postInstall = ''
+    wrapProgram $out/bin/goldwarden \
+      --suffix PATH: ${lib.makeBinPath [dbus pinentry]}
+
+    mkdir -p $out/usr/share/polkit-1/actions/
```
Comment (eclairevoyant) [inline pkgs/by-name/go/goldwarden/package.nix:33]:
```suggestion
```
`install -D` already handles this

File: pkgs/by-name/go/goldwarden/package.nix [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    install -Dm644 $src/resources/com.quexten.goldwarden.policy $out/usr/share/polkit-1/actions/com.quexten.goldwarden.policy
+  '';
+
+  meta = with lib; {
+    description = "A feature-packed Bitwarden compatible desktop integration";
+    homepage = "https://github.com/quexten/goldwarden";
+    license = licenses.mit;
+    maintainers = with maintainers; [ ];
```
Comment (eclairevoyant) [inline pkgs/by-name/go/goldwarden/package.nix:41]: Please add yourself or someone who has confirmed as willing to maintain the package

File: pkgs/by-name/go/goldwarden/package.nix [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  buildInputs = [libfido2];
+
+  postInstall = ''
+    wrapProgram $out/bin/goldwarden \
+      --suffix PATH: ${lib.makeBinPath [dbus pinentry]}
+
+    mkdir -p $out/usr/share/polkit-1/actions/
+    install -Dm644 $src/resources/com.quexten.goldwarden.policy $out/usr/share/polkit-1/actions/com.quexten.goldwarden.policy
+  '';
+
```
Comment (eclairevoyant) [inline pkgs/by-name/go/goldwarden/package.nix:36]: Can we add a `passthru.updateScript` here? Will make it much easier to update. `nix-update-script` is usually a solid option

File: pkgs/by-name/go/goldwarden/package.nix [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    repo = "goldwarden";
+    rev = "v${version}";
+    hash = "sha256-OXJovoJ2+YIMqzoG6J2LlxUC5DMZRAdEl+ZEv6PDXlI=";
+  };
+
+  vendorHash = "sha256-1Px60+f23qoP5eEOUC3WG5vKJYjbD3bPOrDyBpXlMT0=";
+
+  ldflags = [ "-s" "-w" ];
+
+  nativeBuildInputs = [makeWrapper];
```
Comment (eclairevoyant) [inline pkgs/by-name/go/goldwarden/package.nix:25]:
```suggestion
  nativeBuildInputs = [makeBinaryWrapper];
```
preferred when possible, per nixpkgs manual

File: pkgs/by-name/go/goldwarden/package.nix [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+  buildInputs = [libfido2];
+
+  postInstall = ''
+    wrapProgram $out/bin/goldwarden \
+      --suffix PATH: ${lib.makeBinPath [dbus pinentry]}
+
+    mkdir -p $out/usr/share/polkit-1/actions/
+    install -Dm644 $src/resources/com.quexten.goldwarden.policy $out/usr/share/polkit-1/actions/com.quexten.goldwarden.policy
```
Comment (eclairevoyant) [inline pkgs/by-name/go/goldwarden/package.nix:34]:
```suggestion
    install -Dm644 $src/resources/com.quexten.goldwarden.policy -t $out/usr/share/polkit-1/actions
```

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

File: pkgs/by-name/go/goldwarden/package.nix (hunk 1 above)
Comment (arthsmn) [reply pkgs/by-name/go/goldwarden/package.nix:33]: Done.

File: pkgs/by-name/go/goldwarden/package.nix (hunk 2 above)
Comment (arthsmn) [reply pkgs/by-name/go/goldwarden/package.nix:41]: Sorry, forgot. Done.

File: pkgs/by-name/go/goldwarden/package.nix (hunk 4 above)
Comment (arthsmn) [reply pkgs/by-name/go/goldwarden/package.nix:25]: Done.

File: pkgs/by-name/go/goldwarden/package.nix (hunk 5 above)
Comment (arthsmn) [reply pkgs/by-name/go/goldwarden/package.nix:34]: Done.

File: pkgs/by-name/go/goldwarden/package.nix (hunk 3 above)
Comment (arthsmn) [reply pkgs/by-name/go/goldwarden/package.nix:36]: Done.

File: pkgs/by-name/go/goldwarden/package.nix [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+  nativeBuildInputs = [makeBinaryWrapper];
+
+  buildInputs = [libfido2];
+
+  postInstall = ''
+    wrapProgram $out/bin/goldwarden \
+      --suffix PATH : ${lib.makeBinPath [dbus pinentry]}
+
+    install -Dm644 $src/resources/com.quexten.goldwarden.policy -t $out/usr/share/polkit-1/actions
```
Comment (eclairevoyant) [inline pkgs/by-name/go/goldwarden/package.nix:34]:
```suggestion
    install -Dm644 $src/resources/com.quexten.goldwarden.policy -t $out/share/polkit-1/actions
```

whoops, missed this earlier

**Review by eclairevoyant — CHANGES_REQUESTED** [review]

File: pkgs/by-name/go/goldwarden/package.nix (hunk 6 above)
Comment (arthsmn) [reply pkgs/by-name/go/goldwarden/package.nix]: Oh, thanks! Done.

**Review by eclairevoyant — APPROVED** [review]

**Comment by symphorien** [comment]
I think the required system services should be directly be installed in the package or setup by a NixOS module. The imperative `setup` commands don't fit NixOS very well.

**Comment by arthsmn** [comment]
> I think the required system services should be directly be installed in the package or setup by a NixOS module. The imperative `setup` commands don't fit NixOS very well.

I couldn't find the required file, so I think it is generated at runtime. @quexten what do you think? Do you have plans to have a preset file so we can copy from? Do you want to have a hardcoded file or let upstream handle the service? Keep in mind that it would be desirable to have a service in NixOS in case your app gets options to be enabled etc.

**Comment by quexten** [comment]
For the other packaging formats I have the upstream issue https://github.com/quexten/goldwarden/issues/12, which I have not been able to deal with just yet. For now, the service file is a hard-coded string https://github.com/quexten/goldwarden/blob/e15cee5aa92749b395960498dd069b8794025e68/cmd/setup_linux.go#L69-L77, but if it helps you, I can make this a dedicated file upstream that you can just use. This is installed as a user-service.

**Comment by arthsmn** [comment]
> For the other packaging formats I have the upstream issue [quexten/goldwarden#12](https://github.com/quexten/goldwarden/issues/12), which I have not been able to deal with just yet. For now, the service file is a hard-coded string https://github.com/quexten/goldwarden/blob/e15cee5aa92749b395960498dd069b8794025e68/cmd/setup_linux.go#L69-L77, but if it helps you, I can make this a dedicated file upstream that you can just use. This is installed as a user-service.

This would be awesome! Thanks!

**Comment by nomeata** [comment]
I gave this a try, but it would never show a pinentry for me, even after I added pinentry-gtk2 to the system packages. (Maybe I should have restarted or at least re-logged-in?)

But somehow the journal log says 
```
Mär 25 21:21:11 riviera .goldwarden-wrapped[2553526]: [INF] [21:21] [Goldwarden > Pinentry] >>> Asking for pin |Unlock Goldwarden|Enter the vault PIN|
```
without any indication that it couldn’t fin a pinenry or so?

Maybe the package should have a suitable pinentry as a runtime dependency so that it will work for everyone? Not sure.

**Comment by quexten** [comment]
Probably a good idea. Upstream has a GTK based custom pinentry fallback for the UI, which is currently only available as a Flatpak, but @SuperSandro2000 is looking into packaging the full UI version without flatpak for nix which would also have this fallback: #293117

**Comment by nomeata** [comment]
Yeah, just found that PR and saw it installs `config.programs.gnupg.agent.pinentryPackage`. I guess I will wait for that PR.

**Pull request merged**